default/python/AI/pathfinding.py

```diff
@@ -0,0 +1,86 @@
+from heapq import heappush, heappop
+from collections import namedtuple
+
+import freeOrionAIInterface as fo
+from turn_state import state
+
+
+# TODO Allow short idling in system to refuel
+# TODO Consider additional optimizations:
+#    - Check if universe.shortestPath is valid before running pathfinder
+#      (seemingly not worth it - we have exact heuristic in that case)
+#    - Cut off branches that can't reach target due to supply distance
+#    - For large graphs, check existance on a simplified graph (use single node to represent supply clusters)
+#    - For large graphs, check if there are any must-visit nodes (e.g. only possible resupplying system),
+#      then try to find the shortest path between those and start/target.
+def find_path_with_resupply(start, target, fleet_id):
+    path_information = namedtuple('path_information', ['distance', 'fuel', 'path'])
+    universe = fo.getUniverse()
+    empire_id = fo.empireID()
+    fleet = universe.getFleet(fleet_id)
+    supplied_systems = set(fo.getEmpire().fleetSupplyableSystemIDs)
+
+    start_fuel = fleet.maxFuel if start in supplied_systems else fleet.fuel
+
+    # We have 1 free jump from supplied system into unsupplied systems.
+    # Thus, the target system must be at most maxFuel + 1 jumps away
+    # in order to reach the system under standard conditions.
+    # In some edge cases, we may have more supply here than what the
+    # supply graph suggests. For example, we could have recently refueled
+    # in a system that is now blockaded by an enemy or we have found
+    # the refuel special.
+    target_distance_from_supply = -state.get_systems_by_supply_tier(target)
+    if (fleet.maxFuel + 1 < target_distance_from_supply and
+            universe.jumpDistance(start, target) > start_fuel):
+        # can't possibly reach this system
+        return None
+
+    # initialize data structures
+    path_cache = {}
+    shortest_possible_path = universe.shortestPathDistance(start, target)
```

Comment:
I don't see a stated requirement that it has been previously verified that start and target are connected at all (in some cases the AI might be considering a potential target which is itself visible but does not have a known path connection to the galaxy portion the AI inhabits.

Seems to me that we should check here if shortest_possible_path is -1 and return None in that case.

Up above there is a comment referring to testing if universe.shortestPath is valid, and some reference  an exact heuristic, but I am not entirely sure what to make of that comment